src/allocation_engine/interval_tree.rs

```diff
@@ -185,8 +185,173 @@ struct InnerNode {
     right: Option<Box<InnerNode>>,
     /// Cached height of the node.
     height: u64,
-    /// Cached maximum valued covered by this node.
-    max_key: u64,
+}
+
+impl InnerNode {
+    /// Creates a new InnerNode object.
+    #[allow(dead_code)]
+    fn new(key: Range, node_state: NodeState) -> Self {
+        InnerNode {
+            key,
+            node_state,
+            left: None,
+            right: None,
+            height: 1,
+        }
+    }
+
+    /// Returns a readonly reference to the node associated with the `key` or
+    /// None if the searched key does not exist in the tree.
+    #[allow(dead_code)]
+    fn search(&self, key: &Range) -> Option<&InnerNode> {
+        match self.key.cmp(key) {
+            Ordering::Equal => Some(self),
+            Ordering::Less => self.right.as_ref().and_then(|node| node.search(key)),
+            Ordering::Greater => self.left.as_ref().and_then(|node| node.search(key)),
+        }
+    }
+
+    /// Returns a readonly reference to the node associated with the `key` or
+    /// None if there is no Node representing an interval that covers the
+    /// searched key. For a key [a, b], this method will return a node with
+    /// a key [c, d] such that c <= a and b <= d.
+    #[allow(dead_code)]
+    fn search_superset(&self, key: &Range) -> Option<&InnerNode> {
+        if self.key.contains(key) {
+            Some(self)
+        } else if key.end < self.key.start && self.left.is_some() {
```

Comment:
I wonder if we can use the same trick `.as_ref().and_then()` as in `search()` and get rid of unwraps and `else { None }`. Something like
```rust
if self.key.contains(key) {
    Some(self)
} else if key.end < self.key.start {
    self.left.as_ref().and_then(|node| node.search_superset(key))
} else { // seems like the only possibility here is key.start > self.key.end
    self.right.as_ref().and_then(|node| node.search_superset(key))
}
```